Update Navigation component version to 2.4.2

oguzkocer opened the pull request:
This PR updates navigation component version from `2.3.5` to `2.4.2`.

This PR is a prerequisite for the `Plugin DSL` PR #15314. This is because `2.3.5` version of the component is requesting `org.jetbrains.kotlin:kotlin-compiler-embeddable:1.4.31` which creates a conflict with our Kotlin version `1.6.10`. I am not entirely sure why the conflict only happens with the plugin DSL changes, and I am sure there are other ways to address this issue, but upgrading this dependency seemed like the best way forward for us.

I realized the lint issue from the initial commit after I've opened the PR because the task wasn't failing for me locally. I looked into the issue and ended up disabling the new lint check for `NullSafeMutableLiveData` due to bugs within our dependencies. The lint error also makes it clear that the crash is happening within lint checks and not in our application. I've documented my findings and the reason for disabling check as a comment. When the issue within `lifecycle-livedata` is resolved, we can look into re-enabling the check along with fixing 20+ new issues it raises 😱 

Fixing this ^ lint issue resulted in a couple `FragmentLiveDataObserve` lint errors, another new lint check as far as I can tell. Thankfully addressing those in 2d2973d1fef9ff2e4b9be6113f06120d4e81a738 was a lot simpler with a change that's consistent with our codebase.

---

[Looking through the release notes for `2.4.0`](https://developer.android.com/jetpack/androidx/releases/navigation#version_240_3), the only breaking change as far as I can tell is the following:

> All Navigation artifacts have been rewritten in Kotlin. This has improved the nullability of classes using generics (such as NavType subclasses).

The only error we got from this change is due to `outputFileExtension` being nullable in `PreviewImageFragment`. I have no familiarity with the code, so I went with a `null`-check since we already have another one for `R.id.preview_dest`. However, I can see the `outputFileExtension` being optional, so I was hoping @ashiagr could help me with the correct approach here.

> All Kotlin extension functions that used to be part of the -ktx artifacts have been moved to their respective main artifacts. The -ktx artifacts will continue to be published, but are completely empty.

This sounds like another breaking change, however [declaring dependencies section](https://developer.android.com/jetpack/androidx/releases/navigation#declaring_dependencies) still refers to the `-ktx` versions for Kotlin and the dependencies are still being fetched successfully. There is also no instructions in the release notes to update the dependency declaratoin, so I am leaning towards leaving them as is for now.

> Version 2.4.1 & 2.4.2

Version [2.4.1](https://developer.android.com/jetpack/androidx/releases/navigation#version_241_3) & [2.4.2](https://developer.android.com/jetpack/androidx/releases/navigation#version_242_3) only contain bug fixes - which is expected since they are just patches.

### To test:

I have no familiarity with Navigation component or how it's used. So, I just did a smoke test of the PR. It'd be great if someone who is familiar with the component and its usage in our app can do a more detailed testing. I am hoping @ashiagr can help with that. 🙇‍♂️ 

## Regression Notes
1. Potential unintended areas of impact

* Unlikely, but potential breakage of navigation where the navigation component is used - probably in the image editor.
* Possible mishandling of navigation when `outputExtension` is missing while previewing. To address this concern, I asked for help from a more experienced developer to verify it's handled correctly.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
I don't know if we have tests that cover navigation component. I did a smoke test and pinged a more experienced developer to help test the changes.

3. What automated tests I added (or what prevented me from doing so)
Adding new tests is out of the scope for this PR.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

oguzkocer committed:
Update navComponentVersion to 2.4.2

oguzkocer commented:
@ashiagr I converted this PR back to draft due to the lint error. Sorry about that. I'll try to address it as soon as I can.

oguzkocer committed:
Ignore NullSafeMutableLiveData lint error

oguzkocer committed:
Disable NullSafeMutableLiveData from build.gradle

oguzkocer committed:
Fixes FragmentLiveDataObserve lint errors

oguzkocer commented:
@ashiagr I've addressed the lint issue, updated the PR description and marked the PR ready for review. I am sorry about the extra pings.

oguzkocer commented:
@wordpress-mobile/owl-team [The transient dependency version changes](https://github.com/wordpress-mobile/WordPress-Android/pull/16632#issuecomment-1136311588) resulting from this update is a bit worrisome. We are already not using the declared dependencies for most of these dependencies, but it'd have been better if we could update those before we had to update the navigation component. Historically we have turned a blind eye to these transient version updates and I'd really like to change that and that's why I've been working on the dependency catalog infrastructure.

Having said that, practically speaking I think it might makes sense to still merge this PR especially since it's somewhat of a blocker for the Plugin DSL work in #15314. I am saying `somewhat` because we might be able to find an alternative solution to the kotlin metadata issue in that PR, but I don't think it'll be one we'll be too happy about either.

I can see 3 ways forward:

1. Review/test/merge this & Plugin DSL PRs as soon as we can so @ParaskP7 is not blocked with his project.
2. Open separate PRs for the [major version updates](https://github.com/wordpress-mobile/WordPress-Android/pull/16632#issuecomment-1136311588), merge them in and then get back to this PR. This will require us to invest significant time to these updates and without a process in place to keep the dependencies updated, we may need to undertake a similar task soon. This may block @ParaskP7's project. (I don't think it'll be anything more than a merge conflict, but I might be mistaken)
3. Leave this and plugin DSL PRs as a draft and handle them when we can update our dependencies as a whole. (especially androidx ones that depend on each other) The caveat with that is it might take a long time before we get there. We might also end up having a similar conundrum with a different dependency update soon.

I don't have a strong opinion, but if I had to make a decision, I'd stay practical and go with option 1. What do you think?

ParaskP7 commented on libs/image-editor/ImageEditor/src/main/kotlin/org/wordpress/android/imageeditor/preview/PreviewImageFragment.kt:
```diff
@@ -271,14 +271,16 @@ class PreviewImageFragment : Fragment(R.layout.preview_image_fragment) {
-                    ),
-                    navOptions
-            )
+            outputFileExtension?.let {
```
Warning (⚠️): @oguzkocer @ashiagr I tested this update and it seems that when you add an image to a post and then you try to `Edit` it, a toast starts to appear saying `Failed to crop and save image, please try again.`.

I also reviewed the `outputFileExtension` argument and it was evident that it was defined as expected within the navigation graph, which is `nullable`, so this got me by surprise, meaning how come the auto-generated `PreviewImageFragmentDirections.actionPreviewFragmentToCropFragment(...)` action didn't pick that up. 🤔 

After searching a bit, I found this [issue](https://issuetracker.google.com/issues/198689811) within Google's issue tracker and it got me thinking if us adding the `android:defaultValue="@null"` to this `outputFileExtension` argument within the navigation graph will help to both fix the compile and runtime issue. I actually tried that and it seems to work. 🎉 

@ashiagr can you verify that the above problem and potential solution is what you would expect as well? 🙇

ashiagr replied:
Great find, @ParaskP7. I tested the scenario and indeed the issue is resolved by the solution you mentioned. Thanks for linking the Android [issue](https://issuetracker.google.com/issues/198689811). 
Also thanks @oguzkocer for updating the library. 🙇‍♀️ 

We should  
- Remove `null`-check added for `outputFileExtension` in f73798c.
- Add `android:defaultValue="@null"` to `outputFileExtension` argument within the navigation graph for both the fragments: `preview_dest` and `crop_dest`.

@oguzkocer, let me know if it can be handled as part of this PR itself or if I should create another PR with a fix. 🙏

ParaskP7 replied:
Thank for the confirmation and testing this solution @ashiagr ! 🙇 🚀

ParaskP7 replied:
Thank for the confirmation and testing this solution @ashiagr ! 🙇 🚀

oguzkocer replied:
Thanks for the investigation @ParaskP7 and verifying @ashiagr 🙇‍♂️ I made the suggested changes in 51af542de4a64d6a26681cf5c5b2e2131f94e676, hopefully that's correct.

ashiagr replied:
Changes in 51af542 look good to me, thanks @oguzkocer!

ParaskP7 replied:
Thank you for the changes @oguzkocer ! 🚀 🌟

ParaskP7 commented on WordPress/build.gradle:
```diff
@@ -270,6 +270,22 @@ android {
+
+        There are several bug fix reports in ther lifecycle release notes. I can see 5 bug fixes
+        related to it if I search for `NullSafeMutableLiveData`:
+        * https://developer.android.com/jetpack/androidx/releases/lifecycle NullSafeMutableLiveData)
```
Praise (❤️): @oguzkocer This is a good catch about the various bugs with `NullSafeMutableLiveData` and the fact that the latest stable `lifecycle` version of `2.4.1` should address them. It might be good if this would be our next dependency to update.

oguzkocer replied:
@ParaskP7 Unfortunately `lifecycle` `2.4.1` in combination of this navigation component update doesn't seem to fix all the issues we are having. I ran a quick test yesterday and I still had some crashes. Having said that, I agree with updating `lifecycle` as our next dependency and trying to re-enable this check. 🤞 I missed something in my test run and it will work with that update.

ParaskP7 replied:
Yes, exactly, as per my [comment below](https://github.com/wordpress-mobile/WordPress-Android/pull/16632#discussion_r881619708) I don't think that updating `lifecycle` to `2.4.1` will fix all those 35 issues, maybe some of them. I still think that the rest of them are valid errors that we will need to fix anyway.

PS: I am actually not sure why Lint didn't error on those before and it did only after updating `navigation` to `2.4.2`. Maybe those were warnings instead of errors before, maybe this `navigation` update added [transitive dependencies](https://github.com/wordpress-mobile/WordPress-Android/pull/16632#issuecomment-1136311588) that made this `NullSafeMutableLiveData` thing more strict on the Lint side of things, not sure... 🤷

ParaskP7 commented on WordPress/build.gradle:
```diff
@@ -270,6 +270,22 @@ android {
+
+        Disabling this from `lint.xml` file doesn't seem to work, that's why it's added here.
+        */
+        disable "NullSafeMutableLiveData"
```
Suggestion (💡): @oguzkocer Having said [the above](https://github.com/wordpress-mobile/WordPress-Android/pull/16632/files#r881611141) I still believe that this `NullSafeMutableLiveData ` check is a valid check, that produces good results, even with our current `2.2.0` version of `lifecycle`. As such, I wouldn't recommend disabling it at this point.

Out of the 35, I can totally recommend suppressing 10 of them (inline and closer to the source), as they indeed seem false positives to me, but for the rest I recommend fixing them. I actually spend a bit of time on it and the fixes are not that hard. I can share the patch I ended up with if you like.

PS: Maybe the newest `2.4.1` version of the `lifecycle` library would automatically fix those 10 false positives that I noticed. 🤷

oguzkocer replied:
@ParaskP7 If we don't disable the check, lint task crashes due to a bug in `NullSafeMutableLiveData` checks. This is explained in [this comment](https://github.com/wordpress-mobile/WordPress-Android/pull/16632/files#diff-276cc53d974bad3212f8e1f709d32ea3f8d380c4dccb10419505d283fb41f2d7R283-R284):

> We are disabling this check for now as it results in a **crash**, but we should try to enable it again as we update our `lifecycle` version.

oguzkocer replied:
@ParaskP7 I tried to improve the comment in 77940f86cd7641ef3b49c44cc65caea4174b45ba. Let me know if that makes things clearer as to why we have to disable this check at this time.

ParaskP7 replied:
@oguzkocer This is very interesting that lint task crashed on your side, it doesn't crash for me locally, do you mean it crashed on CI maybe? 🤔 Locally for me it completes successfully with a failure because of those 35 new errors. Maybe I am still misunderstanding something in terms of the task crashing... 🤷

ParaskP7 replied:
PS: After removing `disable "NullSafeMutableLiveData"`, I actually managed to run Lint successfully after fixing quickly those 25 `NullSafeMutableLiveData` Lint errors and suppressing 10 of those, without even a failure.

oguzkocer replied:
@ParaskP7 Yes, it's a CI failure, it didn't crash for me locally either. However, my local runs don't catch all the errors/failures, so I don't trust it all that much 😓 Here is the [initial failure](https://buildkite.com/automattic/wordpress-android/builds/4816#c2794969-a9e1-4384-8f27-20b5a3a9444e) which I thought was clear from [this comment](https://github.com/wordpress-mobile/WordPress-Android/pull/16632#issuecomment-1136331951), but I should have still clarified - sorry about that!

If you can branch from this PR and get a successful run in CI, I'd be happy to merge it into this PR. I really wanted to get it to work as well, but since it's a crash in a dependency, I couldn't come up with any ideas about how we can fix it in the project without updating the dependency.

I wonder if we could fix the "lint error" in the mentioned classes 🤔 There is no lint error reported due to the crash, but from the lint check type, we might be able to figure out how to improve the code which could fix the crash as well. Admittedly I didn't think about that yesterday.

ParaskP7 replied:
> @ParaskP7 Yes, it's a CI failure, it didn't crash for me locally either. However, my local runs don't catch all the errors/failures, so I don't trust it all that much 😓 Here is the [initial failure](https://buildkite.com/automattic/wordpress-android/builds/4816#c2794969-a9e1-4384-8f27-20b5a3a9444e) which I thought was clear from https://github.com/wordpress-mobile/WordPress-Android/pull/16632#issuecomment-1136331951, but I should have still clarified - sorry about that!

Thanks for sharing @oguzkocer , yes this [initial failure](https://buildkite.com/automattic/wordpress-android/builds/4816#c2794969-a9e1-4384-8f27-20b5a3a9444e) you shared was the same I was getting locally, lint failed with 37 errors, locally I was getting 34 errors but also 3 generic lint errors (total 38), but those [3 generic "LintError: Lint Failure" ones](https://s3.amazonaws.com/buildkiteartifacts.com/b26f1746-2f0c-49cc-bbd1-a45a5e24c351/e9cd0950-28ae-4b22-8385-c4c19564f99b/0180f760-69c5-4714-970d-43ee98570e9f/c2794969-a9e1-4384-8f27-20b5a3a9444e/WordPress/build/reports/lint-results-wordpressVanillaRelease.html?response-content-type=text%2Fhtml&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAQPCP3C7L2QXFF6U5%2F20220526%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20220526T073835Z&X-Amz-Expires=600&X-Amz-Security-Token=IQoJb3JpZ2luX2VjELX%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJGMEQCIDgOV1Jqxh2HRv9J9PYqgpzSmw0AzZ362SkRuivj8dI9AiAmoWBW0fL%2BJ1Wdgn%2Fa0PnDh9pZPiMN%2B8eQkDecLRfarSrbBAie%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F8BEAAaDDAzMjM3OTcwNTMwMyIMBx3Qyhj2p8YC7IC5Kq8EXg63zVy99K1C44Z7ooYfppr7IPl80MKcEkecX3TFO6RdIgkITHccaL4L75pSIXqhhnEdVv5g6mP2MtODZRy9Zs0j%2F4pqrCfdg%2FlU1oSoRQCNhOKknZ%2FtZcHOwrCXThghHnst%2FKI%2F1WtAv60PI1%2BsACTF5IcRMZP7HXAiJD7U6AdCEaAN2Bmw%2BVCOkQP3oFB%2BqpAxpC%2BpjkLQuWbgKOmNbl1H7hbKeqF4lsgMyQ6pTFl5B7Wcto0aepBTxXFamj3xOp3EPu%2B5eTftNA0V2g%2FlDT0vXlBt3pfabDTLDaWbSLGwqKyNGbYaWwYseh%2BeSdGn0lM7F%2Bx1gphWdM%2B7N6SWqmlMOatFKunn4PBXQcE1quukYJSvG5n6iezuPyE0e7Xp8fmBdcZ9nQtZTybYc88myxxxk%2FO0iiMfEzCtdFvWsCLludRTKVCmal4HkTD%2FZuIIC4tvTMUjvyid6EdWuvBLjobBnc%2FEFyN80SKE8nUP57tEgWsIiXLJKVg9Tt3TbLdefofvB5kn9lIEDbDdWOBMjyCc02E95TpbjUybhb6qj2ToNvPnXsGe%2BJ0CJwpzEOcZiFBNYJBWp%2FerqbApAcwSDiyOUGG8onvEG9wfn8P9Py%2FnOb84tWIZ6GK0eQVA%2Bx13qkw1fxpMzuThtlVKAj2rCAtIzkuCXKLfOd3328HKDzS7Z0%2F2y1YgNo195W%2FI%2F5NohQIESxsjgRTd7dnYWYNmuJWLhufDBwIE0WXjIOFgyzDriryUBjqqAQjobB%2FM%2Bv1y7HpsdMOh4otcCa8tkZ0xQVSR23qhmpOSASv8xJ47NgDPBcvAUO72%2Flq4QxGhx1QMOgjVvyAqnDv4j15gbo2INNtA0xCgAx98T%2BNQ6tbX9lYkpmZ3kflj0GHP2o9EXC%2Bjjq4zN%2FqJs8o1Cfkl8kBuuuEWY5cvi869gd0m1dyAUITs8tI7015vqEizLB5%2BEKrPrSN1IQ5ffCo7SXExyQUlW1Gc&X-Amz-SignedHeaders=host&X-Amz-Signature=d7cf03937e509e8abdb7b48537192aa9af4c6b86d6a36e0be928a926b5e41b5e#LintError) all referred to specific classes within those other more specific 32 `NullSafeMutableLiveData` errors and as such it was evident for me that if those specific errors are fixed, then the generic ones would be probably fixed as well, and luckily enough they were fixed indeed (even though those 2 out of 3 `DiffCallback` generic lint errors were weird ones).

What throw me off was the fact that you mention that the lint is crashing due to the bugs you mentioned. To me, lint crashing vs. lint failing with errors mean separate things and thus I wanted to clarify that. Initial I though that lint couldn't even complete on your side and as the lint task was crashing. But, now I understand what you meant by that. 👍 

> If you can branch from this PR and get a successful run in CI, I'd be happy to merge it into this PR. I really wanted to get it to work as well, but since it's a crash in a dependency, I couldn't come up with any ideas about how we can fix it in the project without updating the dependency.

Yes, let me do that so that we could avoid disabling the `NullSafeMutableLiveData` error altogether for the whole project, which in most case is a good indication of something wrong going on and it would be great if engineers would fix that instead instead of ignoring it, on the spot in their PRs, so that we could avoid merging these issues into our `trunk`.

Please expect a PR from my side today. 👍 

> I wonder if we could fix the "lint error" in the mentioned classes 🤔 There is no lint error reported due to the crash, but from the lint check type, we might be able to figure out how to improve the code which could fix the crash as well. Admittedly I didn't think about that yesterday.

👍 💯

ParaskP7 replied:
> Please expect a PR from my side today. 👍

FYI: The PR is now ready: #16647

oguzkocer replied:
@ParaskP7 It looks like the link task from https://github.com/wordpress-mobile/WordPress-Android/pull/16647 is also crashing.

> To me, lint crashing vs. lint failing with errors mean separate things and thus I wanted to clarify that.

They are objectively separate things and the issue we have is a crash, not a lint failure. This is clear from the lint report from the original commit as well as the new PR you've opened:

```
[../../src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt](https://s3.amazonaws.com/buildkiteartifacts.com/b26f1746-2f0c-49cc-bbd1-a45a5e24c351/e9cd0950-28ae-4b22-8385-c4c19564f99b/018100af-fba3-400f-b620-e682a3f7f814/018100b0-1b18-4561-83e9-e9632eb4668a/WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt): Unexpected failure during lint analysis of BaseStatsUseCase.kt (this is a bug in lint or one of the libraries it depends on)

Stack: ArrayIndexOutOfBoundsException:

You can run with --stacktrace or set environment variable LINT_PRINT_STACKTRACE=true to dump a full stacktrace to stdout.
[../../src/main/java/org/wordpress/android/ui/layoutpicker/LayoutCategoryDiffCallback.kt](https://s3.amazonaws.com/buildkiteartifacts.com/b26f1746-2f0c-49cc-bbd1-a45a5e24c351/e9cd0950-28ae-4b22-8385-c4c19564f99b/018100af-fba3-400f-b620-e682a3f7f814/018100b0-1b18-4561-83e9-e9632eb4668a/WordPress/src/main/java/org/wordpress/android/ui/layoutpicker/LayoutCategoryDiffCallback.kt): Unexpected failure during lint analysis of LayoutCategoryDiffCallback.kt (this is a bug in lint or one of the libraries it depends on)

Message: Index 0 out of bounds for length 0

The crash seems to involve the detector androidx.lifecycle.lint.NonNullableMutableLiveDataDetector.
You can try disabling it with something like this:
    android {
        lintOptions {
            disable "NullSafeMutableLiveData"
        }
    }

Stack: ArrayIndexOutOfBoundsException:NonNullableMutableLiveDataDetector$createUastHandler$1.getFieldTypeReference(NonNullableMutableLiveDataDetector.kt:101)
←NonNullableMutableLiveDataDetector$createUastHandler$1.visitClass(NonNullableMutableLiveDataDetector.kt:88)
←UElementVisitor$DispatchPsiVisitor.visitClass(UElementVisitor.kt:600)
←UElementVisitor$DelegatingPsiVisitor.visitClass(UElementVisitor.kt:1208)
←AbstractKotlinUClass.accept(AbstractKotlinUClass.kt:42)
←ImplementationUtilsKt.acceptList(implementationUtils.kt:23)
←AbstractKotlinUClass.accept(AbstractKotlinUClass.kt:45)
←ImplementationUtilsKt.acceptList(implementationUtils.kt:23)

You can run with --stacktrace or set environment variable LINT_PRINT_STACKTRACE=true to dump a full stacktrace to stdout.
[../../src/main/java/org/wordpress/android/ui/stats/refresh/lists/StatsBlockDiffCallback.kt](https://s3.amazonaws.com/buildkiteartifacts.com/b26f1746-2f0c-49cc-bbd1-a45a5e24c351/e9cd0950-28ae-4b22-8385-c4c19564f99b/018100af-fba3-400f-b620-e682a3f7f814/018100b0-1b18-4561-83e9-e9632eb4668a/WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/StatsBlockDiffCallback.kt): Unexpected failure during lint analysis of StatsBlockDiffCallback.kt (this is a bug in lint or one of the libraries it depends on)

Stack: ArrayIndexOutOfBoundsException:

You can run with --stacktrace or set environment variable LINT_PRINT_STACKTRACE=true to dump a full stacktrace to stdout.
```

ParaskP7 replied:
Oh shoot, this is so very weird, I'll try and investigate further to figure this out, it does run successfully for me locally. Do this branch fails on you locally @oguzkocer ? 🤔

ParaskP7 replied:
@oguzkocer FYI, I think I managed to put this `NullSafeMutableLiveData` Lint error together to understand why it got introduced in the first place as part of this PR and how it might potentially be solved as part of [this accompanying](https://github.com/wordpress-mobile/WordPress-Android/pull/16647) PR I just did.

My understanding below:
- As part of this PR, the transitive `lifecycle` dependency got update to `2.3.1` (from our current `2.2.0`).
- As part of this `2.3.1` lifecycle update and its [initial alpha01](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.3.0-alpha01) version Google introduced a new Lint check that warns you when setting a `null` value on ` MutableLiveData` that has been defined in Kotlin as non-null (see `New Features`).
- However, this newly added `NullSafeMutableLiveData` Lint check is buggy.
- First bug fix attempt with the [2.3.0-alpha06](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.3.0-alpha01) version.
- Second bug fix attempt with the [2.3.0-alpha07](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.3.0-alpha07) version.
- Last bug fix attempt with the [2.4.0-alpha02](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.4.0-alpha02) version.
- As such, this made me think that we would need to update our `lifecycle` dependency to at least the `2.4.0` in an attempt to see if these `NullSafeMutableLiveData` Lint bugs will get resolved.
- The current latest stable `lifecycle` version is the `2.4.1`. It might be better to attempt and update to it instead.
- ❌ I just tried updating `lifecycle` to `2.4.1` with my [latest commit](https://github.com/wordpress-mobile/WordPress-Android/pull/16647/commits/7ee2401bc0a4f6197e314ee153d8ec62e31ab42e) on [my accompanying](https://github.com/wordpress-mobile/WordPress-Android/pull/16647) PR, but it actually didn't help. 😞 
- ⚙️ I'll need to investigate further. Especially the fact that locally Lint succeeds on my side.

ParaskP7 replied:
Omg, it is getting trickier and trickier, I just noticed that we have a duplication of the `lifecycleVersion` version declaration in our root level `build.gradle` file and thus my update above failed as it did nothing, I updated the first declaration of `lifecycleVersion` to `2.4.1` but it got overwritten by the second one, which was again `2.2.0`. 🤷 

I am on it... stay tuned! 😅

ParaskP7 replied:
UPDATE: Ok, I am now making progress and updating to `2.4.1` works. However, it added 10 more `NullSafeMutableLiveData`, which to be honest, from the release notes, it is actually expected. This probably why Lint was failing with those weird errors that got it crashing. 🤞 

PS: I am running out of time today, I'll continue with this tomorrow. I converted the [my accompanying](https://github.com/wordpress-mobile/WordPress-Android/pull/16647) PR back to draft for now.

oguzkocer replied:
@ParaskP7 Thank you for posting your understanding, it's in line with my findings. I tried to explain exactly these in the [code comments from this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/16632/files#diff-276cc53d974bad3212f8e1f709d32ea3f8d380c4dccb10419505d283fb41f2d7R275-R281). Maybe those make more sense now? Let me know if you have suggestions to improve them.

---

Good catch on the double `lifecycleVersion`! That might be the issue I had in my brief try with the updated `lifecycleVersion`. I don't think it changes this PR or the recommendation of trying to re-enable the lint check with the new `lifecycleVersion` when we update it. However, it might change our prioritization of when we'll do that `lifecycleVersion` update.

---

I think we are finally on the same page about the issue. Since this PR and `plugin DSL` are prerequisite to your project, I am happy to leave the decision of whether to merge this as is or address the lint issue first. However, in my opinion, it might be good to time box your work on this. Lint checks are important, but a single lint check being disabled for a few sprints (if necessary) will not have a significant impact on the project - if it did, we'd prioritize fixing those lint warnings in the first place. I think this is especially true for this particular lint check as it has quite a few implementation bugs and doesn't seem mature yet.

oguzkocer replied:
> UPDATE: Ok, I am now making progress and updating to 2.4.1 works. However, it added 10 more NullSafeMutableLiveData, which to be honest, from the release notes, it is actually expected. This probably why Lint was failing with those weird errors that got it crashing. 🤞

Just wanted to note that I saw this comment after I posted my last comment. 🤞 it works!

ParaskP7 replied:
👋 @oguzkocer !

Yes, we are on the same page now! 💯 

I agree with timeboxing the Lint error resolution work I am doing, I think I did the bulk already and it will only take me a couple of additional hours today to complete it (maybe less). So, let ne give this a try today as well and then we can disable this Lint check if the update to `2.4.1` fails as well.

Will keep you posted! 🤞

ParaskP7 replied:
UPDATE: And success, updating `lifecycle` to `2.4.1` and resolving/suppressing those 10 new `NullSafeMutableLiveData` lint errors did the trick. As such, the [accompanying PR](https://github.com/wordpress-mobile/WordPress-Android/pull/16647) is now ready for review again. 🎉 

Cc @oguzkocer @ashiagr

ParaskP7 commented:
👋 @oguzkocer and thanks this update PR! 🥇 

> https://github.com/orgs/wordpress-mobile/teams/owl-team https://github.com/wordpress-mobile/WordPress-Android/pull/16632#issuecomment-1136311588 resulting from this update is a bit worrisome.

True... 🤔 

> We are already not using the declared dependencies for most of these dependencies, but it'd have been better if we could update those before we had to update the navigation component. Historically we have turned a blind eye to these transient version updates and I'd really like to change that and that's why I've been working on the dependency catalog infrastructure.

This is very thoughtful, I agree. 💯 

At some point I want to also bring to our attention this [dependency-analysis-android-gradle-plugin](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin) plugin, which amongst other things it does, it is really good on helping with `transitive dependencies` (see [point.2 here](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin#detect-unused-and-misused-dependencies)). I personally have been using this plugin for a year now, I have seen positive feedback from the community on it and I can only recommend us using it going forward (it also went [stable recently](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/main/CHANGELOG.md#version-100)).

> Having said that, practically speaking I think it might makes sense to still merge this PR especially since it's somewhat of a blocker for the Plugin DSL work in https://github.com/wordpress-mobile/WordPress-Android/pull/15314. I am saying somewhat because we might be able to find an alternative solution to the kotlin metadata issue in that PR, but I don't think it'll be one we'll be too happy about either.

I personally agree. 👍 

> 1. Review/test/merge this & Plugin DSL PRs as soon as we can so @ParaskP7 is not blocked with his project.

I agree. We should just be really thorough with our testing before merging this PR with `trunk`. 👍 

> 2. Open separate PRs for the https://github.com/wordpress-mobile/WordPress-Android/pull/16632#issuecomment-1136311588, merge them in and then get back to this PR. This will require us to invest significant time to these updates and without a process in place to keep the dependencies updated, we may need to undertake a similar task soon. This may block @ParaskP7's project. (I don't think it'll be anything more than a merge conflict, but I might be mistaken)

We can do that too, but it depends on how we would like to proceed going forward and with which projects. 🤔 

> Leave this and plugin DSL PRs as a draft and handle them when we can update our dependencies as a whole. (especially androidx ones that depend on each other) The caveat with that is it might take a long time before we get there. We might also end up having a similar conundrum with a different dependency update soon.

I would personally take the risk and proceed with the [Plugin DSL](https://github.com/wordpress-mobile/WordPress-Android/pull/15314/files) PR and all the work needed to be done to help with that (like this PR) and not keep pushing this change for later. However, I do understand if we decide to do it the other way around. 🤔 

> I don't have a strong opinion, but if I had to make a decision, I'd stay practical and go with option 1. What do you think?

I am with you on that. I would choose option 1. too. 👍

oguzkocer committed:
Adds defaultValue @null to outputFileExtension in nav_graph

oguzkocer committed:
Improve the comment for disabling NullSafeMutableLiveData lint check

ParaskP7 committed:
Analysis: Remove global disabling of null safe mutable live data

There are currently 32 errors related to 'NullSafeMutableLiveData'. All
those will get resolved or suppressed locally in subsequent commits.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for base list use case

The 'mutableSnackbarMessage' mutable live data value is set to null
during 'onCleared()' and as such it needs to be declared as nullable.

Then, using 'let' on the 'snackbarMessage' live data this can be guarded
against so that to avoid a null pointer exception.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for show error message

The 'showErrorMessage' mutable live data value is set to
'event.error.message' on multiple occasions. However, 'event.error' can
be nullable and as such the 'event.error.message' itself.

There is an 'if (event.isError) { ... }' statement above to guard from
that ever happening, but the compiler is not smart enough to understand
that. Maybe with 'Kotlin Contracts' this problem would have been
avoided, but this feature is a separate discussion of its own.

Thus, by guarding this with 'let' before setting this 'showErrorMessage'
mutable live data is the best that can be done at this point of time.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for supported states

The 'supportedStates' mutable live data value is set to null during
'onCountrySelected(...)' and can be also set to null via the
'OnDomainSupportedStatesFetched.supportedStates' from within the
'SiteStore' as it is actually defined as nullable. As such, the
'supportedStates' itself needs to be declared as nullable.

ParaskP7 committed:
Refactor: Optimize import for feature announcement

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for feature announcement

The 'currentFeatureAnnouncement' mutable live data can be set to null
during 'loadFeatures()' and via
'FeatureAnnouncementProvider.getLatestFeatureAnnouncement(...)' which
might produce a nullable 'FeatureAnnouncement'. As such, the
'currentFeatureAnnouncement' itself needs to be declared as nullable.

The rest of the changes are propagated due to the above change and are
mainly guarding the rest of the functionality that depends on that with
'let' and '?' scope functions.

ParaskP7 committed:
Refactor: Remove unused live data extensions (from 2017)

This 'LiveDataExtensions' is not used anywhere within the codebase and
as such can be safely remove. If at any point in the future it will be
needed again, engineers should either bring it back, with a more model
version of it, or use another utility from the 'LiveData' library
itself, which will serve the same purpose.

This also solves two 'NullSafeMutableLiveData' lint errors related to
this function.

ParaskP7 committed:
Refactor: Suppress null safe mutable live data for live data utils

Suppress this 'NullSafeMutableLiveData' instead of changing this
'LiveData.mapAsync(...)' extension function and as such its signature
seems to be the safest choice at the moment since this extension
function is being utilized in extension places.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for media picker

The 'selectedIds' mutable live data can be set to null during the
'start(...)' and via its 'selectedIds' parameter which is nullable. As
such, the 'selectedIds' itself needs to be declared as nullable.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for photo picker

The 'selectedIds' mutable live data can be set to null during the
'start(...)' and via its 'selectedIds' parameter which is nullable. As
such, the 'selectedIds' itself needs to be declared as nullable.

Although the initial logic was different and what I am describing above,
and setting the 'selectedIds' was guarded with 'let' I chose to update
this logic and align that with the 'MediaPickerViewModel' logic and the
update I did there with my previous commit. Actually, this makes sense
as this view model and its associated fragment is treating this
'selectedIds' live data value as nullable already and guarding against
it accordingly.

ParaskP7 committed:
Refactor: Optimize import for plans view model

ParaskP7 committed:
Refactor: Remove unused cached plans live data

The live data 'cachedPlans' is only utilized from within the view model
and as such can be replaced with the '_cachedPlans' mutable live data.

This change will make 'cachedPlans' intention much more straightforward.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for plans

The 'plans' and 'cachedPlans' mutable live data value can be set to null
during 'onPlanOffersFetched(...)' because of the
'OnPlanOffersFetched.planOffers' from within the
'PlansOfferStore' as it is actually defined as nullable. As such, the
'plans' and 'cachedPlans' themselves need to be declared as nullable.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for plugin browser

Defaulting to an empty title is enough for this use case.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for post list

The 'searchQuery' mutable live data is set to null during
'clearSearch()'. As such, the 'searchQuery' itself needs to be declared
as nullable.

Also, this view model's associated fragment and activity screens are
treating this 'searchQuery' live data value as nullable already and use
it as such accordingly.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for add category

The 'navigateBack' mutable live data can be set to null during
'cleanupAndFinish()'. As such, the 'navigateBack' itself needs to be
declared as nullable.

Also, this view model's associated fragment is treating this
'navigateBack' live data value as nullable already and guards against it
accordingly.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for publish settings

The 'onNotificationTime' mutable live data can be set to null during
'onNotificationCreated()'. As such, the 'onNotificationTime' itself
needs to be declared as nullable.

Also, this view model's associated fragment is treating this
'onNotificationTime' live data value as nullable already and guards
against it accordingly.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for selected date provider

The 'selectedDateChanged' mutable live data is set to null during the
'clear(...)' functions. As such, the 'selectedDateChanged' itself
needs to be declared as nullable.

Also, this view model's associated fragments are treating this
'selectedDateChanged' live data value as nullable already and guard
against it accordingly.

ParaskP7 committed:
Refactor: Optimize import for site creation intents view model

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for site creation

The 'onIntentSelected' mutable live data can be set to null during
'onCustomVerticalSelected()' because of the 'IntentsUiState.searchQuery'
as it is actually defined as nullable. As such, the 'onIntentSelected'
itself needs to be declared as nullable.

However, this view model's associated activity, main view model and
interface were expecting a non-nullable intent. But, this can actually
change easily and the interface intent can be defined and used as
nullable as well. It ends us anyway being using within the
'SiteCreationMainVM.onSiteIntentSelected(...)', which can accept a
nullable intent without any problems.

ParaskP7 committed:
Refactor: Rename mutable site changed to expected backing field syntax

ParaskP7 committed:
Analysis: Suppress null safe mutable live data for stats site provider

Suppress this 'NullSafeMutableLiveData' instead of changing this
'LiveData.observeEvent(...)' extension function and as such its
signature seems to be the safest choice at the moment since this
extension function is being utilized in extension places.

For more info see 'StatsFragment' and the usage of
'viewModel.siteChanged.observeEvent(...)' which prevents us from
changing this easily.

ParaskP7 committed:
Refactor: Rename mutable snackbar message to a backing field syntax

Also, this change moves this 'showSnackbarMessage' backing field closer
to it equivalent and associated public live data field.

ParaskP7 committed:
Analysis: Suppress null safe mutable live data for stats view all

Suppress this 'NullSafeMutableLiveData' instead of changing this
'LiveData.observeEvent(...)' extension function and as such its
signature seems to be the safest choice at the moment since this
extension function is being utilized in extension places.

For more info see 'StatsViewAllFragment' and the usage of
'viewModel.showSnackbarMessage.observeEvent(...)' which prevents us from
changing this easily.

ParaskP7 committed:
Refactor: Remove redundant let call from on subfilter page updated

ParaskP7 committed:
Analysis: Suppress false positive null safe mutable live data errors

For these classes in this commit the 'NullSafeMutableLiveData' Lint
error is suppressed are false positive. It is false positive because the
nullability of the variable is already checked before it is assigned to
the mutable live data field, but it is anyway ignored as being
potentially null and thus reported.

ParaskP7 committed:
Build: Update lifecycle to 2.4.1

ParaskP7 committed:
Build: Remove duplicate lifecycle declaration and update to 2.4.1

As part of this update the below occurred:
- The 'lifecycle-extensions' artifact and APIs got removed as it has
been deprecated since '2.2.0' and it is no longer updated. The
recommendation nowadays is to instead add dependencies for the specific
lifecycle artifacts needed.
- Update 'ViewModelProvider.Factory.create(...)' signature to a non-null
'ViewModel' generic as it is now defined as such by the Java
factory interface.

The above were the only breaking change. Apart from that everything else
seems to be compiling and working as expected (did a quick smoke test).
However, this update added 10 new 'NullSafeMutableLiveData' lint errors
that will be fixed in subsequent commits.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for base list use case

The 'data' mediator live data is set to null during the 'onCleared()'.
As such, the 'data' itself needs to be declared as nullable. This use
case's associated view model is updated accordingly.

Also, this view model's associated fragment is treating this
'uiModel' live data value as nullable already.

ParaskP7 committed:
Refactor: Optimize import for page parent and search view model

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for page parent search

The 'searchPages' mutable live data is set to null during
'clearSearch()'. As such, the 'searchPages' itself needs to be declared
as nullable.

Also, this view model's associated search view model is treating this
'searchPages' live data value as nullable already and guarding against
it accordingly.

ParaskP7 committed:
Analysis: Resolve null safe mutable live data for pages

The 'searchPages' mutable live data is set to null during
'clearSearch()'. As such, the 'searchPages' itself needs to be declared
as nullable.

Also, this view model's associated search view model is treating this
'searchPages' live data value as nullable already and guarding against
it accordingly.

ParaskP7 committed:
Analysis: Suppress null safe mutable live data for stats screens

Suppress all 'NullSafeMutableLiveData' instead of changing this
'mergeNotNull(...)' implementation and as such 'showSnackbarMessage'
field's signature. This seems to be the safest choice for now.

ParaskP7 committed:
Analysis: Suppress null safe mutable live data for pages

Suppress this 'NullSafeMutableLiveData' instead of changing this
'MutableLiveData.postOnUi(...)' extension function and as such its
signature seems to be the safest choice at the moment since this
extension function is being utilized in extension places.

AliSoftware commented:
Moving this PR's milestone to `20.1` as I'm doing `20.0` code freeze today.

oguzkocer committed:
Merge pull request #16647 from wordpress-mobile/lint/resolve-null-safe-mutable-live-data-lint-errors

Resolve `NullSafeMutableLiveData` Lint Errors

oguzkocer commented:
@ParaskP7 @ashiagr I just merged #16647 into this PR, so this is ready for another look.

ParaskP7 commented:
👋 @oguzkocer !

> @ParaskP7 @ashiagr I just merged https://github.com/wordpress-mobile/WordPress-Android/pull/16647 into this PR, so this is ready for another look.

Great, thank you so much, I am on it! 🏃 🥇

ParaskP7 reviewed (approved):
👋 @oguzkocer !

I reviewed and tested this PR again, this time even more thoroughly with my general smoke testing and my more focused image editor testing, everything looks good and I didn't encounter any hiccups.

As per our conversation [here](https://github.com/wordpress-mobile/WordPress-Android/pull/16632#issuecomment-1137217543) and us staying practical and go with option 1, I agree, let's merge this to `trunk` and then update the [Plugin DSL](https://github.com/wordpress-mobile/WordPress-Android/pull/15314/files) PR accordingly to move forward.

As far as the transitive dependencies are concerned, I took a closer look at the [dependency tree diff](https://github.com/wordpress-mobile/WordPress-Android/pull/16632#issuecomment-1142330100) and the "main" dependencies I see this PR affecting is the below:
- [Lifecycle](https://developer.android.com/jetpack/androidx/releases/lifecycle)
- [Fragment](https://developer.android.com/jetpack/androidx/releases/fragment)

-----

For the [Lifecycle](https://developer.android.com/jetpack/androidx/releases/lifecycle) part, we already handled that with more care, as part of [this accompanying](https://github.com/wordpress-mobile/WordPress-Android/pull/16647) PR of mine, and then reviewed it in more detail.

-----

For the [Fragment](https://developer.android.com/jetpack/androidx/releases/fragment) part I can note the below:
- I am seeing on most occasions that the transitive fragment `1.3.6` dependency is being updated to `1.4.1` for the `com.automattic.about`, `com.automattic.stories`, `androidx.legacy`, `androidx.preference`, `androidx.work`, `com.google.android.gms`, `com.google.dagger`, `com.zendesk` libraries and our `libs:editor` module.
- This means to me that the changes we should expect are:
  - [Version 1.4.0](https://developer.android.com/jetpack/androidx/releases/fragment#1.4.0) related, which as far as I can tell is:
    - `FragmentStrictMode`, new API that is not affecting us.
    - `FragmentContainerView`, new API that is not affecting us.
    - `FragmentScenario`, new implementation that is not affecting us.
    - `FragmentTransaction`, new transaction that is not affecting us.
    - `FragmentManager`, new experimental API that is not affecting us.
    - `Multiple back stacks` 🥳 ...but not affecting us.
  - [Version 1.4.1](https://developer.android.com/jetpack/androidx/releases/fragment#1.4.1) related, which as far as I can tell is:
    - `Bug Fixes`, which are good to have on top of `1.4.0` anyway, thus not affecting us.

-----

As such, I am comfortable enough that this change will not break our app, at least to the extend that we wouldn't be able to recover fast afterwards. @ashiagr it would be great if I get a second opinion on that too in case I am missing something vital.

PS: In addition to that and since we already did the [Fragment](https://developer.android.com/jetpack/androidx/releases/fragment) update work, in terms of verifying everything works as expected in our app, we could create another PR right after this one to explicitly update out current outdate `1.2.4` fragment version to `1.4.1`. I'll can create an issue on that right now so that we don't forget (done #16678).

oguzkocer commented:
Thanks for the investigation for the transitive dependency updates @ParaskP7! I agree that with `lifecycle` updated, remaining updates are less of a risk.

Since @ashiagr reviewed both this PR as well as the #16647, I am going to merge this PR now to keep things moving for @ParaskP7's project. However, @ashiagr if you have any other feedback, please let me know and I'll prioritize opening a new PR to address it.
